### PR TITLE
needsPackage updates

### PR DESCRIPTION
**Pull request by d-torrance**
Two small updates:

* Have `packageFiles User` return the empty list, which fixes the current problem of `needsPackage User` raising a key error (startup.m2 isn't in the loaded files hash table since it's a string not a file)
* Add a warning message about recreating instances of types when reloading a package

Closes: #3851

### CODE REVIEW

File: M2/Macaulay2/m2/packages.m2 [hunk 1]
```diff
@@ -553,6 +555,7 @@ debug GlobalDictionary := dict -> (
     checkShadow())
 
 packageFiles = pkg -> (
+    if pkg#"pkgname" == "User" then return {};
```
Comment (mahrud) [inline M2/Macaulay2/m2/packages.m2]: Is this the only package that is problematic?
How about a package defined directly in the interpreter?

Comment (d-torrance) [reply M2/Macaulay2/m2/packages.m2]: Good point!

```m2
i2 : newPackage "Foo"

o2 = Foo

o2 : Package

i3 : endPackage "Foo"

o3 = Foo

o3 : Package

i4 : packageFiles oo

o4 = {/home/profzoom/src/macaulay2/M2/M2/Macaulay2/m2/stdio}

o4 : List
```

I'll play around with this some more

Comment (d-torrance) [reply M2/Macaulay2/m2/packages.m2]: This should be addressed now

File: M2/Macaulay2/m2/packages.m2 [hunk 2]
```diff
@@ -553,10 +555,15 @@ debug GlobalDictionary := dict -> (
     checkShadow())
 
 packageFiles = pkg -> (
-    pkgaux := (
+    srcfile := realpath pkg#"source file";
+    if (
+	srcfile == "stdio"         or
+	srcfile == "currentString" or
+	match("/startup\\.m2(?:\\.in)?$", srcfile))
```
Comment (mahrud) [inline M2/Macaulay2/m2/packages.m2:562]: why the extra `(?:\\.in)?$`?

File: M2/Macaulay2/tests/normal/testing.m2 [hunk 3]
```diff
@@ -21,6 +21,9 @@ expectedCode = DIV{
 assert BinaryOperation(symbol ===, code pkgtest, expectedCode)
 assert BinaryOperation(symbol ===, code 0, expectedCode)
 
+TEST "assert Equation(1 + 1, 2)"
+check User
```
Comment (mahrud) [inline M2/Macaulay2/tests/normal/testing.m2:25]: Check still calls `needsPackage`, right? does `needsPackage`/`loadPackage` do anything for User?

File: M2/Macaulay2/m2/packages.m2 (hunk 2 above)
Comment (d-torrance) [reply M2/Macaulay2/m2/packages.m2:562]: The file location of `startupString` is startup.m2.in in the build directory by default.  I've patched the Debian package so that it points to the distributed copy of startup.m2 for reproducibility, so this catches both cases.

If you'd like, I could change this regex to just `/startup\\.m2\\.in$` and add a corresponding patch to the Debian package.

File: M2/Macaulay2/tests/normal/testing.m2 (hunk 3 above)
Comment (d-torrance) [reply M2/Macaulay2/tests/normal/testing.m2:25]: Good question!

It's already marked as loaded since #3483, so nothing unless we pass `Reload => true`.  Which currently raises an error.  Same with packages we create in the REPL:

```m2
i1 : loadPackage(User, Reload => true)
 -- warning: reloading User; recreate instances of types from this package
stdio:1:11:(3): error: file not found on path: "User.m2"

i2 : newPackage "Foo"; endPackage "Foo"

o3 = Foo

o3 : Package

i4 : loadPackage(oo, Reload => true)
 -- warning: reloading Foo; recreate instances of types from this package
stdio:3:11:(3): error: file not found on path: "Foo.m2"
```

I'll play around with this some more.

File: M2/Macaulay2/m2/packages.m2 (hunk 2 above)
Comment (mahrud) [reply M2/Macaulay2/m2/packages.m2:562]: I'm saying either way "/startup.m2" should match, no?

Comment (d-torrance) [reply M2/Macaulay2/m2/packages.m2:562]: Oh sure.  I guess I was trying to avoid the very minor edge case of an actual package file whose path contains the string "startup.m2" somewhere in the middle, say `/foo/bar/startup.m2/Foo.m2`.

File: M2/Macaulay2/tests/normal/testing.m2 (hunk 3 above)
Comment (d-torrance) [reply M2/Macaulay2/tests/normal/testing.m2:25]: Now that I think about this some more, maybe the current behavior is the correct one?  We can't reload a package if there's no file to reload it from...

**Review by mahrud — APPROVED** [review]
